earth_enterprise/src/common/FileAccessorPluginLoader.cpp

```diff
@@ -0,0 +1,108 @@
+        return;
+    }
+
+    boost::filesystem::directory_iterator end;
+    std::transform(start, end, std::back_inserter(files), 
+        [pluginDir](const boost::filesystem::directory_entry& entry) {
+            return pluginDir + entry.path().leaf().string(); });
+}
+
+FileAccessorPluginLoader& FileAccessorPluginLoader::Get() {
+    static FileAccessorPluginLoader loader;
+    return loader;
+}
+
+void FileAccessorPluginLoader::DefaultLoadPluginsImpl (const std::string &pluginDirectory, std::vector<PluginFactory> &factories) {
+    assert(factories.size() == 0);
+    
+    std::vector<std::string> files;
+    ScanPluginDirectory(pluginDirectory, files);
+    for (auto pluginFile : files) {
+        FileAccessorPluginLoader::PluginHandle handle = dlopen(pluginFile.c_str(), RTLD_NOW);
+        if (handle){
+            const std::string getFactoryFunctionName = "get_factory_v1";
+            get_factory_t get_factory = reinterpret_cast<get_factory_t>(dlsym(handle, getFactoryFunctionName.c_str()));
+            FileAccessorFactory *pFactory = get_factory != nullptr ? get_factory() : nullptr;
+            if(pFactory){
+                factories.push_back({handle,pFactory});
+            }
+            else{
+                // pluginFile was a valid shared library, but did not provide the required function
+                dlclose(handle);
+                notify(NFY_WARN, "File accessor plugin %s does not implement the expected function %s", pluginFile.c_str(), getFactoryFunctionName.c_str());
+            }
+        }
+        else {
+            // pluginFile was something other than a shared library
+            notify(NFY_WARN, "Unable to open file accessor plugin %s", dlerror());
+        }
+    }
+}
+
+FileAccessorPluginLoader::FileAccessorPluginLoader(
+        LoadPluginFunc LoadImpl,      // default: nullptr
+        UnloadPluginFunc UnloadImpl,  // default: nullptr
+        std::string pluginDirectory   // default: "/opt/google/plugin/fileaccessor/"
+        ):
+        pluginDir(pluginDirectory) {
+    loadPluginFunc = LoadImpl ? LoadImpl : DefaultLoadPluginsImpl;
+    unloadPluginFunc = UnloadImpl ? UnloadImpl : [](PluginHandle handle){dlclose(handle);};
+}
+
+FileAccessorPluginLoader::~FileAccessorPluginLoader() {
+    for (auto factory : factories){
+        unloadPluginFunc(factory.first);
+    }
+}
+
+void FileAccessorPluginLoader::LoadPlugins() {
+    std::lock_guard<std::mutex> lock(loadMutex);
+    if (loaded)
+        return;
+
+    loadPluginFunc(pluginDir, factories);
+
+    loaded = true;
+}
+
+std::unique_ptr<AbstractFileAccessor> FileAccessorPluginLoader::GetAccessor(std::string file_path) {
+    // Even though LoadPlugins checks `loaded`, check here to avoid incurring
+    // the cost of the lock every time someone asks for a file accessor.
+    if (!loaded)
+        LoadPlugins();
+
+    for(PluginFactory f : factories){
```

Comment:
Minor: as an optimization, you could assume that all files will use the same FileAccessor type, so you can cache the Factory that works for the first file and always try it first on the other files. I'm not sure if that's really worth the effort, though.

Reply:
I wasn't sure if we would always be able to assume that the same file accessor type would always be needed. Is that a safe assumption?

If not, what if we cache a pointer to the last retrieved and try that one first on subsequent calls?

Reply:
As it stands, you can't change the publish root without restarting server, so based on that I would expect that we would always need the same kind of accessor. I guess you could have a hybrid publish root with some files local and some remote, but we could just state that that is not supported. However, the method you suggest is kind of the best of both worlds - cache the last accessor that worked and always try that first, then scan through the other accessors if needed. Again, I'm not sure it's worth the effort, it's just an optimization that came to mind.

Reply:
As I started looking at it, there are a couple of wrinkles with optimizing in the ways described. Adding a variable for the last matching factory would, I think, necessitate adding a mutex around much of the code in `FileAccessorPluginLoader::GetAccessor` to make sure that variable isn't changing out from under you. I don't know what the overhead of mutex locking is, but it would cut into the time savings at least a bit.

As far as the idea of using the same factory each time, there's a bit of a rub with the current implementation: a `POSIXFileAccessorFactory` would need to be implemented. That's not too bad, but it would need to be able to parse a POSIX file path, whereas now a `POSIXFileAccessor` is returned by default if no other suitable plugin was found. 
It's doable, and not too bad, but it's not entirely trivial either. 

I have a slight preference toward punting on that, but I'm open to discussion. I may also be missing an easier way of doing it.

Reply:
I'm OK with punting. It will likely be rare to have more than 1 or 2 plugins installed anyway, so it may not save much time.